blockstore/splitstore/splitstore.go

```diff
@@ -551,6 +591,15 @@ func (s *SplitStore) HeadChange(_, apply []*types.TipSet) error {
 
 	if !atomic.CompareAndSwapInt32(&s.compacting, 0, 1) {
 		// we are currently compacting -- protect the new tipset(s)
+
+		// race: we call beginTxnProtect _after_ setting compacting.
+		// 1. HeadChange triggers a compaction but stops immediately after setting this
+		//    flag.
+		// 2. HeadChange tries ties to trigger another compaction, sees one is already in
```

Comment:
wait, that's not possible. HeadChange notifications are synchronous, we can't have a second one before the first one has returned.

Reply:
I think I will add a comment about this, so as to make the invariant assumption explicit.

Note that this is very unlikely to change, as a lot of things depend on this and all hell will break loose if it ever changes.

Reply:
Ok, fair enough. I'd still feel better with a lock in case someone decides to change that, but I guess it's a reasonable constraint on HeadChange.

Reply:
well, it's not bigie to put a lock just for head changes! Just pointing out it's not necessary, but it will take 30s to put one.

Reply:
done in https://github.com/filecoin-project/lotus/pull/6474/commits/60212c86cbd855ba531fb9e7a8303cf57eb25b67